### PR TITLE
New Worker Abstraction (V2)

### PR DESCRIPTION
**Pull request by jessicafalk**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
second take on a new worker abstraction. Instead of putting wrappers all around it, we remove absolutely everything that is not necessary.
The worker is now a simple class storing some data regarding connection, origin, .. etc, which makes it possible to use it either in monobehaviors or in systems (using the `WorkerConfig`component)
It is tighly coupled to the connection i.e. we connect, when the worker is created, and upon disposing the worker, we will disconnect.

Removed loads of things
moved the translation unit stuff into its own file to make it easier to delete everything later (instead of moving it into the worker which was the initial plan)

This PR is huge: reading it in commit order should make your life a lot easier

#### Tests
ran it locally. it worked ™️

#### Documentation
will come in the next pr

#### Primary reviewers
@jamiebrynes7 @samiwh

### CODE REVIEW

File: code_generator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt [hunk 1]
```diff
@@ -135,19 +135,19 @@ namespace <#= qualifiedNamespace #>
                 <#= componentDetails.CamelCaseTypeName#>.DirtyBit = false;
 
 <# if (componentDetails.IsBlittable) { #>
-                worker.EntityManager.AddComponentData(entity, <#= componentDetails.CamelCaseTypeName#>);
+                EntityManager.AddComponentData(entity, <#= componentDetails.CamelCaseTypeName#>);
 <# } else { #>
-                 worker.EntityManager.SetComponentObject(entity, <#= componentDetails.CamelCaseTypeName #>);
+                 EntityManager.SetComponentObject(entity, <#= componentDetails.CamelCaseTypeName #>);
```
Comment (martinzlocha) [inline code_generator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt]: Nit: Extra space.

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            return worker;
+        }
+
+        public static void Disconnect(Worker worker)
+        {
+            OnDisconnect(worker);
+            worker.Dispose();
+        }
+
+        public void Dispose()
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: Might be better to make this private. Currently its possible to call `worker.Dispose()` or `Worker.Disconnect(worker)`, but only in one case will you get OnDisconnect called. Also this way it better mirrors the initializer.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
-            worker = WorkerRegistry.GetWorkerForWorld(World);
+        private TranslationUnityRegistry translationUnityRegistry;
 
+        protected override void OnCreateManager(int capacity)
+        {
+            base.OnCreateManager(capacity);
+            translationUnityRegistry = TranslationUnityRegistry.WorldToTranslationUnit[World];
             dispatcher = new Dispatcher();
-            SetupDispatcherHandlers();
+            SetupDispatcherHandlers();
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Nit: Extra whitespace.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs [hunk 4]
```diff
@@ -1,33 +1,41 @@
...
+        private Worker worker;
+
         protected override void OnUpdate()
         {
-            if (worker.Connection == null)
-            {
-                return;
-            }
-
+            worker = data.WorkerConfigs[0].Worker;
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Even though it hasn't been our standard so far can we check that the array only contains 1 entry.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs [hunk 5]
```diff
@@ -40,18 +48,52 @@ protected override void OnUpdate()
...
+            worker.EntityMapping.Add(entityId, entity);
         }
 
         private void OnRemoveEntity(RemoveEntityOp op)
         {
-            worker.RemoveEntity(op.EntityId.Id);
+            var entityId = op.EntityId.Id;
+            if (!worker.TryGetEntity(entityId, out var entity))
+            {
+                worker.LogDispatcher.HandleLog(LogType.Error, new LogEvent("Tried to delete an entity but there is no entity associated with that EntityId.")
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Duplicate error strings, could extract into a separate variable.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs [hunk 6]
```diff
@@ -40,18 +48,52 @@ protected override void OnUpdate()
...
+                    .WithField(LoggingUtils.LoggerName, LoggerName)
+                    .WithField(LoggingUtils.EntityId, entityId));
+                return;
+            }
+
+            EntityManager.DestroyEntity(entity);
+            worker.EntityMapping.Remove(entityId);
         }
 
         private void OnDisconnect(DisconnectOp op)
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Should this also call worker.OnDisconnect so that the monobehaviours get a callback as well?

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs [hunk 7]
```diff
@@ -60,8 +102,8 @@ private void SetupDispatcherHandlers()
             dispatcher.OnRemoveEntity(OnRemoveEntity);
             dispatcher.OnDisconnect(OnDisconnect);
             dispatcher.OnCriticalSection(op => { inCriticalSection = op.InCriticalSection; });
-
-            foreach (var translationUnit in worker.TranslationUnits.Values)
+
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Nit: Extra whitespace.

File: workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs [hunk 8]
```diff
@@ -42,5 +42,10 @@ public EntityGameObjectLinker(World world, ILogDispatcher logDispatcher)
             spatialOSComponent.SpatialEntityId = spatialEntityId;
             spatialOSComponent.World = world;
         }
+
+        public void Dispose()
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/EntityGameObjectLinker.cs]: Is this needed given that its never used?

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs [hunk 9]
```diff
@@ -19,20 +20,25 @@ public struct Data
...
 
+        [Inject] private WorkerData workerData;
+
+        private static Vector3 speed = new Vector3(2, 0, 0);
+
         protected override void OnUpdate()
         {
+            if (workerData.Length == 0)
+            {
+                new LoggingDispatcher().HandleLog(LogType.Error, new LogEvent("This system should not have been run without a worker entity"));
```
Comment (martinzlocha) [inline workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: Can this use the default dispatcher?

File: workers/unity/Assets/Playground/Config/ArchetypeConfig.cs [hunk 10]
```diff
@@ -13,15 +13,15 @@ public static class ArchetypeConfig
             WorkerTypeToArchetypeNameToComponentTypes = new Dictionary<string, Dictionary<string, ComponentType[]>>
             {
                 {
-                    UnityGameLogic.WorkerType,
+                    "UnityGameLogic",
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Assets/Playground/Config/ArchetypeConfig.cs]: Agree entirely - we should define a set of constants _somewhere_ for these as they will be used in multiple places

File: workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs [hunk 11]
```diff
@@ -32,50 +32,68 @@ public struct RemovedEntitiesData
...
 
         protected override void OnUpdate()
         {
+            if (workerData.Length == 0)
+            {
+                new LoggingDispatcher().HandleLog(LogType.Error, new LogEvent("This system should not have been run without a worker entity"));
+            }
+
+            var worker = workerData.WorkerConfigs[0].Worker;
+            if (viewCommandBuffer == null)
```
Comment (zeroZshadow) [inline workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs]: This is bad, is there no way to do this in `OnCreateManager` ?

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 12]
```diff
@@ -17,64 +13,35 @@ public class Bootstrap : MonoBehaviour
...
-                connectionConfig.UseExternalIp = workerConfigurations.UseExternalIp;
-#endif
+                    WorkerType = "UnityClient",
+                };
+                CreateWorker(config, Vector3.zero);
+                config = new ReceptionistConfig
+                {
+                    WorkerType = "UnityGameLogic",
+                };
+                CreateWorker(config, new Vector3(10, 0, 0));
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Inline the creation of the configs here -

`CreateWorker(new ReceptionistConfig("UnityClient"), Vector3.zero)`

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 13]
```diff
@@ -140,9 +82,19 @@ public static void DomainUnloadShutdown()
             ScriptBehaviourUpdateOrder.UpdatePlayerLoop();
         }
 
-        private void LoadLevel(WorkerBase worker)
+        private void CreateWorker(ConnectionConfig config, Vector3 origin)
         {
-            Instantiate(Level, worker.Origin, Quaternion.identity);
+            if (config.WorkerType.Equals("UnityGameLogic"))
+            {
+                PlayerLifecycleConfig.CreatePlayerEntityTemplate = PlayerTemplate.CreatePlayerEntityTemplate;
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Why does this need to be conditional? If its set on the client, does it matter? 

Related - does this even belong here?

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs [hunk 14]
```diff
@@ -19,20 +20,25 @@ public struct Data
...
+            }
+
+            var origin = workerData.WorkerConfigs[0].Worker.Origin;
+
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: This boilerplate makes me sad

File: workers/unity/Assets/Playground/Scripts/DisconnectSystem.cs [hunk 15]
```diff
@@ -11,17 +11,20 @@ internal class DisconnectSystem : ComponentSystem
...
         }
 
         [Inject] private DisconnectData data;
 
         protected override void OnUpdate()
         {
-            Debug.LogWarningFormat("Diconnected from SpatialOS with reason: \"{0}\"",
-                data.DisconnectMessage[0].ReasonForDisconnect);
+            data.Configs[0].Worker.LogDispatcher.HandleLog(
+                LogType.Warning,
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/DisconnectSystem.cs]: Does this need to be a warning? Disconnecting is not abnormal flow

File: workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs [hunk 16]
```diff
@@ -16,18 +17,18 @@ public class CleanReactiveComponentsSystem : ComponentSystem
         // Here to prevent adding an action for the same type multiple times
         private readonly HashSet<Type> typesToRemove = new HashSet<Type>();
 
-        private WorkerBase worker;
-
+        private TranslationUnityRegistry translationUnityRegistry;
+
```
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs]: Whitespace 🏳️

File: workers/unity/Assets/Playground/Scripts/EntityInitialization/ArchetypeInitializationSystem.cs [hunk 17]
```diff
@@ -26,58 +26,64 @@ public struct Data
...
+                viewCommandBuffer = new ViewCommandBuffer(worker.LogDispatcher);
+            }
             for (var i = 0; i < data.Length; i++)
             {
                 var archetypeName = data.ArchetypeComponents[i].ArchetypeName;
                 var entity = data.Entities[i];
 
                 ComponentType[] componentTypesToAdd;
-                if (!ArchetypeConfig.WorkerTypeToArchetypeNameToComponentTypes[workerType]
+                if (!ArchetypeConfig.WorkerTypeToArchetypeNameToComponentTypes[worker.WorkerType]
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/EntityInitialization/ArchetypeInitializationSystem.cs]: Driveby - use `out var`

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs (hunk 3 above)
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Whitespace ⚪️

File: workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs [hunk 18]
```diff
@@ -32,50 +32,68 @@ public struct RemovedEntitiesData
             public SubtractiveComponent<GameObjectReference> NoGameObjectReference;
         }
 
+        public struct WorkerData
```
Comment (jamiebrynes7) [inline workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs]: It might be worth pulling this out into core given that its use everywhere as a common data struct.

Then no need to redefine in every system

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs (hunk 7 above)
Comment (zeroZshadow) [inline workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Whitespace ⬜️

File: workers/unity/Packages/com.improbable.gdk.core/Systems/CustomSpatialOSSendSystem.cs [hunk 19]
```diff
@@ -13,18 +14,22 @@ public abstract class CustomSpatialOSSendSystem<T> : ComponentSystem where T : I
...
         {
             base.OnCreateManager(capacity);
 
-            worker = WorkerRegistry.GetWorkerForWorld(World);
-
             spatialOSSendSystem = World.GetOrCreateManager<SpatialOSSendSystem>();
             if (!spatialOSSendSystem.TryRegisterCustomReplicationSystem(typeof(T)))
             {
-                worker.LogDispatcher.HandleLog(LogType.Error, new LogEvent(CustomReplicationSystemAlreadyExists)
+                new LoggingDispatcher().HandleLog(LogType.Error, new LogEvent(CustomReplicationSystemAlreadyExists)
```
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/CustomSpatialOSSendSystem.cs]: You could add it as another generic on the system - but that doesn't feel good necessarily.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs (hunk 5 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: Yup - magic strings should be const

File: workers/unity/Packages/com.improbable.gdk.core/Systems/CustomSpatialOSSendSystem.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/CustomSpatialOSSendSystem.cs]: one idea that came up in between would be to let the systems own the dispatcher instead of the worker as this is more aligned to how logging dispatchers in other projects. However, in this case it is actually possible (after realizing that you actually have access to the injected groups from the very start) and I'll change that

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs (hunk 2 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: can't make that private as it's inheriting from `IDisposable`, but I can move the call into `Dispose`

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs (hunk 4 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: yeah, so far I only checked the length, if necessary (because we have multiple injection groups in a system). If we only have one injection group, then that system will only run if there are any worker entities, but I can add a check for whether there is only one entry

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs (hunk 13 above)
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: it was only set for the gamelogic worker in the previous code, simply keeping the same logic here. 
The overall idea would be to factor it out into it's own monobehavior/gameobject, but I didn't want to make the PR even bigger

File: workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs (hunk 6 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/SpatialOSReceiveSystem.cs]: I think that the `Worker.OnDisconnect` should only be called from the worker class. we don't actually disconnect here, but only later. if we call it later, we might end up calling it twice.
once here and a second time during the actual disconnect.

The alternative would be to only call it here and not when we dispose  / disconnect

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs [hunk 20]
```diff
@@ -19,20 +20,25 @@ public struct Data
 
         [Inject] private Data data;
 
-        private static Vector3 speed = new Vector3(2, 0, 0);
-
-        private Vector3 origin;
-
-        protected override void OnCreateManager(int capacity)
+        public struct WorkerData
```
[the remark this answers is not in the document]
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: it doesn't seem like we have a convention for that. After checking a few files we usually make them `public`. Would prefer to leave it like this to have it the same everywhere and then create a new PR to make it private everywhere. Otherwise this PR will just become even messier :P

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs (hunk 9 above)
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: discussed offline: we don't have the worker here, so we need to create a logging dispatcher

File: workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs (hunk 11 above)
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/EntityInitialization/GameObjectInitializationSystem.cs]: found out that we actually can do it. Thought we don't have access to the injected groups in `OnCreateManager`. will revert that change

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs (hunk 14 above)
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: me too :(
maybe we need to create a `SpatialSystem` after all that takes care of that boilerplate code.

File: workers/unity/Assets/Playground/Scripts/DisconnectSystem.cs (hunk 15 above)
Comment (jessicafalk) [reply workers/unity/Assets/Playground/Scripts/DisconnectSystem.cs]: I guess it depends on why you got disconnected. I simply converted it from `LogWarning` to our dispatcher, but can change it to Log if preferred. I don't have a strong opinion about this

**Comment by jessicafalk** [comment]
just as a note for everyone: The worker class will be changed to inherit from `ScriptpableBehaviourManager ` to make the injection bit easier.
but before that, I'll rebase on top of the experimental core csharp branch and then create a new PR with all these changes in it

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+        public readonly Vector3 Origin;
+        public readonly string WorkerType;
+        public readonly string WorkerId;
+        public readonly World World;
+        public readonly Connection Connection;
+        public readonly ILogDispatcher LogDispatcher;
+
+        // callbacks for monobehaviour
+        public static Action<Worker> OnConnect;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: Make this an event rather than an action. (Backed by a list of actions)

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+
+            World = new World(WorkerId);
+            var entityManager = World.GetOrCreateManager<EntityManager>();
+
+            var workerConfig = new WorkerConfig
+            {
+                Worker = this
+            };
+
+            FindTranslationUnits();
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: Why is this here?

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+            ConnectionUtility.Disconnect(Connection);
+            Connection.Dispose();
+        }
+        
+        
+        // can be deleted soonish
+        public readonly Dictionary<int, ComponentTranslation> TranslationUnits =
+            new Dictionary<int, ComponentTranslation>();
+        
+        public Action<Entity, long> AddAllCommandRequestSenders;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: This should be in a core system on a worker not part of the worker itself.

File: workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public static Worker Connect(ConnectionConfig config, ILogDispatcher logDispatcher, Vector3 origin)
+        {
+            return new Worker(config, logDispatcher, origin);
+        }
+
+        public static void Disconnect(Worker worker)
+        {
+            worker.Dispose();
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Worker/Worker.cs]: This probably isn't what we want. We might want a tick to run your cleanup code with a disconnect message.

File: workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs (hunk 9 above)
Comment (samiwh) [reply workers/unity/Assets/Playground/Scripts/Cubes/CubeMovementSystem.cs]: This kind of error message isn't very useful to a user. Can be explicit about what the error is but we should probably also add something which suggests that it's because they might have started the worker incorrectly.

**Comment by jessicafalk** [comment]
closing it in favor of the newest worker abstraction PR
